Array bounds warning building Host with GCC 8

See #1692 for initial problem report.

The issue is with the `LOAD_PSTR` macro in `FakePgmSpace.h`, which copies data from flash to RAM in 4-byte words. Byte-copying from flash is very inefficient so this is an optimisation for the Esp8266.

This issue has arised now we can compile the code using more up-to-date tools. From https://gcc.gnu.org/gcc-8/changes.html:

`The -Warray-bounds option has been improved to detect more instances of out-of-bounds array indices and pointer offsets. For example, negative or excessive indices into flexible array members and string literals are detected.`

This is the output from the pre-processor for the offending code:

```
  case EVENT_STAMODE_CONNECTED:
  (__extension__({
    static const char log_string[] = "%u " "connect to ssid %s, channel %d" "\r\n";
    char fmtbuf[(((sizeof(log_string)) + 3) & ~3)] __attribute__((aligned(4)));
    memcpy(fmtbuf, log_string, sizeof(fmtbuf));;
    m_printf(fmtbuf, system_get_time(), evt->event_info.connected.ssid, evt->event_info.connected.channel);
  }));
```

We're attempting to do a `memcpy` using the size of the allocated buffer, which is longer than the actual data stored in flash. That doesn't matter on the Esp8266 as we know it's safe to read at least to the next word boundary, but it's not good practice.

my branch is  'origin/fix/LOAD_PSTR_overrun'.
Those compiler errors are gone, but unfortunately there are some new ones:
Libraries/RCSwitch/RCSwitch.cpp: In member function ‘char* RCSwitch::getCodeWordB(int, int, boolean)’:
Libraries/RCSwitch/RCSwitch.cpp:288:12: error: invalid conversion from ‘char’ to ‘char*’ [-fpermissive]
     return '\0';
            ^~~~
Libraries/RCSwitch/RCSwitch.cpp: In member function ‘char* RCSwitch::getCodeWordC(char, int, int, boolean)’:
Libraries/RCSwitch/RCSwitch.cpp:351:12: error: invalid conversion from ‘char’ to ‘char*’ [-fpermissive]
     return '\0';
            ^~~~
Libraries/RCSwitch/RCSwitch.cpp: In member function ‘char* RCSwitch::getCodeWordD(char, int, boolean)’:
Libraries/RCSwitch/RCSwitch.cpp:421:20: error: invalid conversion from ‘char’ to ‘char*’ [-fpermissive]
             return '\0';
                    ^~~~
Libraries/RCSwitch/RCSwitch.cpp:440:20: error: invalid conversion from ‘char’ to ‘char*’ [-fpermissive]
             return '\0';

Plus also I got 
In file included from Libraries/Adafruit_GFX/Adafruit_GFX.cpp:38:
Libraries/Adafruit_GFX/glcdfont.c:10: error: "PROGMEM" redefined [-Werror]
  #define PROGMEM
 
In file included from System/include/debug_progmem.h:20,
                 from Arch/Host/Components/esp_hal/include/esp_systemapi.h:43,
                 from Arch/Host/System/include/user_config.h:7,
                 from Wiring/Arduino.h:5,
                 from Libraries/Adafruit_GFX/Adafruit_GFX.h:5,
                 from Libraries/Adafruit_GFX/Adafruit_GFX.cpp:34:
Wiring/FakePgmSpace.h:150: note: this is the location of the previous definition
 #define PROGMEM __attribute__((aligned(4)))

so some more patches needed for the Libraries it would seem

If I workaround past those compiler errors, the buid goes on to give me this:

C+ Libraries/OneWire/OneWire.cpp
C+ Libraries/APA102/apa102.cpp
AR out/build/Host/Linux/libsming.a
mkdir -p out/build/Host/Linux/Arch/Host/Components/lwip
CMake Error: The source directory "/Arch/Host/Components/lwip/Linux" does not exist.

>> CMake Error: The source directory "/Arch/Host/Components/lwip/Linux" does not exist.

@frankdownunder If you're still getting this error, could you do:

```
make submodules-clean
make lwip V=1
```

Should highlight what's going on

> I've got a suspicion you have ...

He for sure does not have `STRICT=1` set. Without it the default behaviour is to treat re-definition as an error. 
```
Libraries/Adafruit_GFX/glcdfont.c:10: error: "PROGMEM" redefined [-Werror]
#define PROGMEM
```

--- a/Sming/Libraries/Adafruit_GFX/glcdfont.c
+++ b/Sming/Libraries/Adafruit_GFX/glcdfont.c
@@ -6,7 +6,7 @@
  #include <avr/pgmspace.h>
 #elif defined(__ESP8266_EX__)
  #include <WiringFrameworkDependencies.h>
-#else
+#elif !defined(PROGMEM)
  #define PROGMEM
 #endif
  

--- a/Sming/Wiring/FakePgmSpace.cpp
+++ b/Sming/Wiring/FakePgmSpace.cpp
@@ -7,7 +7,7 @@ void *memcpy_P(void *dest, const void *src_P, size_t length)
 {
 	// Yes, it seems dest must also be aligned
 	if (IS_ALIGNED(dest) && IS_ALIGNED(src_P) && IS_ALIGNED(length))
-		return memcpy_aligned(dest, src_P, length);
+		return memcpy(dest, src_P, length);
 
 	auto dest0 = reinterpret_cast<char*>(dest);
 	auto src0 = reinterpret_cast<const char*>(src_P);
@@ -122,19 +122,16 @@ char* strcat_P(char* dest, const char* src_P)
 /** @brief copy memory aligned to word boundaries
  *  @param dst
  *  @param src
- *  @param len
- *  @note dst, src and len must all be aligned to word (4-byte) boundaries
+ *  @param len Size of the source data
+ *  @note dst and src must be aligned to word (4-byte) boundaries
+ *  `len` will be rounded up to the nearest word boundary, so the dst
+ *  buffer MUST be large enough for this.
  *
  */
 void* memcpy_aligned(void* dst, const void* src, unsigned len)
 {
-	assert(IS_ALIGNED(dst) && IS_ALIGNED(src) && IS_ALIGNED(len));
-
-	auto pd = (uint32_t*)dst;
-	auto ps = (const uint32_t*)src;
-	auto n = len / 4;
-	while (n--)
-		*pd++ = *ps++;
+	assert(IS_ALIGNED(dst) && IS_ALIGNED(src));
+	memcpy(dst, src, ALIGNUP(len));
 	return dst;
 }
 

--- a/Sming/Wiring/FakePgmSpace.cpp
+++ b/Sming/Wiring/FakePgmSpace.cpp
@@ -114,6 +114,9 @@ char* strcat_P(char* dest, const char* src_P)
 	return dest;
 }
 
+#endif /* ICACHE_FLASH */
+
+
 /*
  * We implement aligned versions of some system functions to be used strictly on
  * data that is word (4-byte) aligned on all parameters.
@@ -158,5 +161,3 @@ int memcmp_aligned(const void* ptr1, const void* ptr2, unsigned len)
 	auto tail2 = pgm_read_dword(reinterpret_cast<const uint8_t*>(ptr2) + len_aligned);
 	return memcmp(&tail1, &tail2, len - len_aligned);
 }
-
-#endif

--- a/Sming/Wiring/FakePgmSpace.cpp
+++ b/Sming/Wiring/FakePgmSpace.cpp
@@ -139,25 +139,24 @@ void* memcpy_aligned(void* dst, const void* src, unsigned len)
  *  @param p1
  *  @param p2
  *  @param len
- *  @retval int 0 if all bytes match, 1 if they don't
- *  @note p1, p2 and len must all be aligned to word (4-byte) boundaries
- *
- *  Note that unlike the standard memcmp routine we do not perform
- *  a relative comparison (i.e. greater/less than).
+ *  @retval int 0 if all bytes match
+ *  @note p1 and p2 must all be aligned to word (4-byte) boundaries
+ *  len is rounded up to the nearest word boundary
  */
 int memcmp_aligned(const void* ptr1, const void* ptr2, unsigned len)
 {
-	assert(IS_ALIGNED(ptr1) && IS_ALIGNED(ptr2) && IS_ALIGNED(len));
+	assert(IS_ALIGNED(ptr1) && IS_ALIGNED(ptr2));
 
-	auto p1 = (const uint32_t*)ptr1;
-	auto p2 = (const uint32_t*)ptr2;
-	auto n = len / 4;
-	while (n--)
-		if (*p1++ != *p2++)
-			return 1; // Match fail
+	unsigned len_aligned = ALIGNDOWN(len);
+	int res = memcmp(ptr1, ptr2, len_aligned);
+	if(res != 0 || len == len_aligned) {
+		return res;
+	}
 
-	// Match
-	return 0;
+	// Compare the remaining bytes
+	auto tail1 = pgm_read_dword(reinterpret_cast<const uint8_t*>(ptr1) + len_aligned);
+	auto tail2 = pgm_read_dword(reinterpret_cast<const uint8_t*>(ptr2) + len_aligned);
+	return memcmp(&tail1, &tail2, len - len_aligned);
 }
 
 #endif

--- a/Sming/Wiring/FakePgmSpace.h
+++ b/Sming/Wiring/FakePgmSpace.h
@@ -236,7 +236,7 @@ extern "C"
  */
 #define LOAD_PSTR(_name, _flash_str)                                                                                   \
 	char _name[ALIGNUP(sizeof(_flash_str))] __attribute__((aligned(4)));                                               \
-	memcpy_aligned(_name, _flash_str, sizeof(_name));
+	memcpy_aligned(_name, _flash_str, sizeof(_flash_str));
 
 #define _FLOAD(_pstr)                                                                                                  \
 	(__extension__({                                                                                                   \

--- a/Sming/Wiring/FakePgmSpace.h
+++ b/Sming/Wiring/FakePgmSpace.h
@@ -55,27 +55,6 @@ typedef uint32_t prog_uint32_t;
 #define PROGMEM __attribute__((aligned(4))) __attribute__((section(".irom.text")))
 #endif
 
-/*
- * Define and use a flash string inline
- */
-#define PSTR(_str)                                                                                                     \
-	(__extension__({                                                                                                   \
-		DEFINE_PSTR_LOCAL(__c, _str);                                                                                  \
-		&__c[0];                                                                                                       \
-	}))
-
-/*
- * Declare and use a flash string inline.
- * Returns a pointer to a stack-allocated buffer of the precise size required.
- */
-#define _F(_str)                                                                                                       \
-	(__extension__({                                                                                                   \
-		DEFINE_PSTR_LOCAL(_flash_str, _str);                                                                           \
-		LOAD_PSTR(_buf, _flash_str);                                                                                   \
-		_buf;                                                                                                          \
-	}))
-
-
 // flash memory must be read using 32 bit aligned addresses else a processor exception will be triggered
 // order within the 32 bit values are
 // --------------
@@ -120,8 +99,6 @@ static inline uint16_t pgm_read_word_inlined(const void* addr)
 extern "C"
 {
 #endif
-	void* memcpy_aligned(void* dst, const void* src, unsigned len);
-	int memcmp_aligned(const void* ptr1, const void* ptr2, unsigned len);
 	void *memcpy_P(void *dest, const void *src_P, size_t length);
 	size_t strlen_P(const char * src_P);
 	char *strcpy_P(char * dest, const char * src_P);
@@ -170,18 +147,13 @@ extern "C"
 
 #else /* ICACHE_FLASH */
 
-#define PROGMEM
-
-#define PSTR(_str) (_str)
-#define _F(_str) (_str)
+#define PROGMEM __attribute__((aligned(4)))
 
 #define pgm_read_byte(addr) (*(const unsigned char *)(addr))
 #define pgm_read_word(addr) (*(const unsigned short *)(addr))
 #define pgm_read_dword(addr) (*(const unsigned long *)(addr))
 #define pgm_read_float(addr) (*(const float *)(addr))
 
-#define memcpy_aligned(dst, src, len) memcpy(dst, src, len)
-#define memcmp_aligned(ptr1, ptr2, len) memcmp(ptr1, ptr2, len)
 #define memcpy_P(dest, src, num) memcpy((dest), (src), (num))
 #define strlen_P(a) strlen((a))
 #define strcpy_P(dest, src) strcpy((dest), (src))
@@ -196,6 +168,27 @@ extern "C"
 
 #endif /* ICACHE_FLASH */
 
+/*
+ * Define and use a flash string inline
+ */
+#define PSTR(_str)                                                                                                     \
+	(__extension__({                                                                                                   \
+		DEFINE_PSTR_LOCAL(__c, _str);                                                                                  \
+		&__c[0];                                                                                                       \
+	}))
+
+/*
+ * Declare and use a flash string inline.
+ * Returns a pointer to a stack-allocated buffer of the precise size required.
+ */
+#define _F(_str)                                                                                                       \
+	(__extension__({                                                                                                   \
+		DEFINE_PSTR_LOCAL(_flash_str, _str);                                                                           \
+		LOAD_PSTR(_buf, _flash_str);                                                                                   \
+		_buf;                                                                                                          \
+	}))
+
+
 #define pgm_read_byte_near(addr) pgm_read_byte(addr)
 #define pgm_read_word_near(addr) pgm_read_word(addr)
 #define pgm_read_dword_near(addr) pgm_read_dword(addr)
@@ -204,6 +197,9 @@ extern "C"
 #define pgm_read_word_far(addr) pgm_read_word(addr)
 #define pgm_read_dword_far(addr) pgm_read_dword(addr)
 #define pgm_read_float_far(addr) pgm_read_float(addr)
+
+void* memcpy_aligned(void* dst, const void* src, unsigned len);
+int memcmp_aligned(const void* ptr1, const void* ptr2, unsigned len);
 
 /** @brief define a PSTR
  *  @param _name name of string

--- a/Sming/Wiring/FlashString.h
+++ b/Sming/Wiring/FlashString.h
@@ -144,7 +144,7 @@
  */
 #define LOAD_FSTR(_name, _fstr)                                                                                        \
 	char _name[(_fstr).size()] __attribute__((aligned(4)));                                                            \
-	memcpy_aligned(_name, (_fstr).data(), sizeof(_name));
+	memcpy_aligned(_name, (_fstr).data(), (_fstr).length());
 
 /*
  * Define a flash string and load it into a named char[] buffer on the stack.

--- a/Sming/Wiring/FlashString.h
+++ b/Sming/Wiring/FlashString.h
@@ -241,7 +241,7 @@ struct FlashString {
 			return false;
 		if(flashData == str.flashData)
 			return true;
-		return memcmp_aligned(flashData, str.flashData, ALIGNUP(flashLength)) == 0;
+		return memcmp_aligned(flashData, str.flashData, flashLength) == 0;
 	}
 
 	bool isEqual(const String& str) const

--- a/Sming/Wiring/WString.cpp
+++ b/Sming/Wiring/WString.cpp
@@ -211,7 +211,7 @@ String &String::copy(flash_string_t pstr, unsigned int length)
 	}
 	else
 	{
-		memcpy_aligned(buffer, (PGM_P)pstr, length_aligned);
+		memcpy_aligned(buffer, (PGM_P)pstr, length);
 		buffer[length] = '\0';
 		len = length;
 	}

